### ISSUE TITLE
Variable from Literal definition changes type in class

### ISSUE THREAD

**Issue by karolyi** [microsoft/pyright#3729]
**Describe the bug**
Hi,

I have a variable that I pass into a class with a certain type. It seems that pyright converts it down to the lowest denominated compatible type?

**To Reproduce**
```python
from typing import Literal

FlavorType = Literal[1, 2, 3]

class Test(object):

    def __init__(self, y: FlavorType):
        self.yy = y

    def test(self):
        print(self.yy)

```

**Expected behavior**
`self.yy` in the test() should stay `FlavorType`, not `int`.

**Screenshots or Code**
Pointing on `self.yy` in `__init__()`:
![image](https://user-images.githubusercontent.com/987055/180262292-37d7ef3a-efb7-4a4a-9b5b-1c413461af8a.png)
Pointing on `self.yy` in `test()`:
![image](https://user-images.githubusercontent.com/987055/180262356-6152ec23-5d7a-4583-883b-ea6302c0d029.png)

**VS Code extension or command-line**
I'm using the latest pyright (automatically updates itself on each relaunch to latest) in Sublime Text with Sublime-LSP

**Additional context**
The above code only works with defining `self.yy: FlavorType = y` in `__init__()`, whereas it should infer its type as `FlavorType`.

**Comment by karolyi** [comment]
Now why is this an issue you might ask?

Here's why:
```python
FlavorType = Literal[1, 2, 3]

VAL_1 = 1
VAL_2 = 2
VAL_3 = 3
TEST_DICT = dict(((VAL_1, 'a'), (VAL_2, 'b'), (VAL_3, 'c')))


class Test(object):

    def __init__(self, y: FlavorType):
        self.yy = y

    def test(self):
        print(TEST_DICT[self.yy])
```
Pointing on `TEST_DICT`:
![image](https://user-images.githubusercontent.com/987055/180264666-dfcc5408-5961-4cd4-acb2-7cd9dd92c5fd.png)
Pointing on the print statement (that is now marked as bad):
![image](https://user-images.githubusercontent.com/987055/180264812-0e07a2b0-25bd-42eb-bcc1-292130ecd5ac.png)

**Comment by erictraut** [comment]
This is intended behavior. The type of instance variable `yy` is not declared, so its type is inferred from assignments using normal inference rules. Literal types are not retained as part of the normal inference rules. After all, if we modified your sample to be `self.yy = 0`, you wouldn't expect the type of `yy` to be inferred as `Literal[0]`; you would expect it to be inferred as `int`. If you want `yy` to be constrained to a literal type, you need to declare its type as such.

**Comment by karolyi** [comment]
Aha.

I can live with that, it's only weird if you look at the examples I linked. Defining the dict that way will make pyright infer the key and value types, instead of going with the same behavior as it did with the class variable.

I can force the class variable type, yes, but this example is still valid as a unintended side-effect of this behavior maybe. Defining dicts like this is valid, and I use these in my django projects to look up field choice display values in a fast way.

**Comment by erictraut** [comment]
The type of the `dict` in your example is dictated by the way the constructor of `dict` is defined in `builtins.pyi`. Pyright is simply applying the normal type rules for function calls there. 

If you were to define your `dict` using a dictionary expression, then pyright's will apply normal inferencing rules.

```python
TEST_DICT = { VAL_1: "a", VAL_2: "b", VAL_3: "c" }
reveal_type(TEST_DICT) # dict[str, int]
```

**Comment by karolyi** [comment]
Thanks, one always learns something new.

As said, the django way to define choice-value pairs is tuple, and it's fastest to pass them to dicts to be able to look up the values later on, hence this way of defining them.

I will use the type specification in my classes.
